### PR TITLE
Add DataMapping annotation and new configuration

### LINKED ISSUE #19: [roq-data] Generated bindings to data file
The goal is to have the records generated some way. TBD

### PR DESCRIPTION
**Pull request by mcruzdev**
Fixes #19

### CODE REVIEW

File: roq-data/deployment/src/main/java/io/quarkiverse/roq/data/deployment/RoqDataConfig.java [hunk 1]
```diff
@@ -19,10 +20,17 @@ public interface RoqDataConfig {
...
+     * with the data from the file.
+     */
+    @WithDefault("false")
+    boolean enforceBean();
+
     static boolean isEqual(RoqDataConfig q1, RoqDataConfig q2) {
-        if (!Objects.equals(q1.dir(), q2.dir())) {
-            return false;
-        }
-        return true;
```
Comment (melloware) [inline roq-data/deployment/src/main/java/io/quarkiverse/roq/data/deployment/RoqDataConfig.java:26]: Don't you also need to check enforceBean in the isEqual?  Typically we compare all the props.

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by mcruzdev** [comment]
I will squash it after because I mentioned a commit in Zulip.

File: roq-data/deployment/src/main/java/io/quarkiverse/roq/data/deployment/RoqDataProcessor.java [hunk 2]
```diff
@@ -54,4 +58,47 @@ void generateInjectable(BuildProducer<SyntheticBeanBuildItem> beansProducer,
...
+                Object object = jsonObject.mapTo(clazz);
+                beansProducer.produce(SyntheticBeanBuildItem.configure(clazz)
+                        .scope(roqDataMapping.isRecord() ? Singleton.class : ApplicationScoped.class)
+                        .addQualifier().annotation(Named.class).addValue("value", roqDataMapping.getName()).done()
+                        .setRuntimeInit()
+                        .runtimeValue(roqDataRecorder.createRoqDataJson(object))
+                        .done());
+            } else if (roqDataMapping.getData() instanceof JsonArray array) {
+
+                if (roqDataMapping.isRecord()) {
```
Comment (mathecruz) [inline roq-data/deployment/src/main/java/io/quarkiverse/roq/data/deployment/RoqDataProcessor.java]: @ia3andy @mkouba

**Comment by ia3andy** [comment]
@mcruzdev is it ready for review,

**Comment by mcruzdev** [comment]
> @mcruzdev is it ready for review?
> 
> 

I stopped because we need definition about how the user will use the data mapping feature.

File: roq-data/deployment/src/main/java/io/quarkiverse/roq/data/deployment/ReadRoqDataProcessor.java [hunk 3]
```diff
@@ -42,6 +54,57 @@ void scanDataFiles(RoqProjectBuildItem roqProject, RoqDataConfig config, BuildPr
...
+    private List<String> checkConfigurationErrors(Map<String, String> annotationMap, Map<String, Object> dataJsonMap) {
+        List<String> messages = new ArrayList<>();
+        for (String name : annotationMap.keySet()) {
+            if (!dataJsonMap.containsKey(name)) {
+                messages.add("The @DataMapping#value('%s') does not match with any data file".formatted(name));
+            }
+        }
+        for (String name : dataJsonMap.keySet()) {
+            if (!annotationMap.containsKey(name)) {
+                messages.add("The data file '%s' does not match with any data mapping".formatted(name));
```
Comment (melloware) [inline roq-data/deployment/src/main/java/io/quarkiverse/roq/data/deployment/ReadRoqDataProcessor.java]:
```suggestion
                messages.add("The data file '%s' does not match with any @DataMapping class".formatted(name));
```

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by mcruzdev** [comment]
sorry @ia3andy @melloware  I'm still on vacation. But now we can use parent mapping and `Instance<Type>` mapping.

I think you can take a initial looking.

**Review by melloware — APPROVED** [review]

**Comment by ia3andy** [comment]
I will give it a try today

**Review by ia3andy — CHANGES_REQUESTED** [review]
Great stuff!!! Thanks @mcruzdev you roq!

I am getting ClassNotFound exception on a trial local project. I think you need to use Jandex to locate the class:
https://quarkus.io/guides/writing-extensions#scanning-deployments-using-jandex


Could you also add the doc section about data (just add a new page with an index like we do on web-bundler or quinoa, we'll add the rest of the content late)? -- if we don't start it it will be too much to catch up..

**Comment by ia3andy** [comment]
image

**Review by ia3andy — APPROVED** [review]

**Comment by ia3andy** [comment]
@mcruzdev add the doc in another PR ;-)

**Pull request merged**